### PR TITLE
Hyperbahn client

### PR DESCRIPTION
**Pull request by ShanniLi**
r @willsalz @truncs @anson627 @junchaowu 

In this PR, tchannel-java can advertise to Hyperbahn and serve requests.
Next steps:
1. TChannel SubChannel: allow connection pooling for different services.
2. Retries

### CODE REVIEW

File: tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+        advertiseTimer.schedule(new TimerTask() {
+            @Override
+            public void run() {
+                try {
+                    ListenableFuture<Response<AdvertiseResponse>> responseFuture = advertise();
+                } catch (Exception ex) {
+                    logger.error(service + " failed to advertise. " + ex.getMessage());
+                }
+            }
+        }, advertiseInterval);
```
Comment (truncs) [inline tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java]: Wasn't the requirement to not have a fixed advertiseInterval and change the internval depending if the advertise response is a success or a failure ?

File: tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+        } catch (TChannelError ex) {
+            // TODO: should be moved into tchanne as retry ...
+            this.logger.error("Advertise failure: " + ex.toString());
+
+            // re-throw for now
+            // TODO: should really handle internally
+            throw ex;
+        }
+
+        responseFuture.addListener(new Runnable() {
```
Comment (truncs) [inline tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java]: In my initial implementation I was using daemon threads which have a nice priority of having a lower priority as compared to other threads and they they terminate themselves with the JVM

File: tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+        this.logger.info("HyperbahnClient shutdown complete.");
+    }
+
+    public static class Builder {
+
+        private final String service;
+        private final TChannel channel;
+
+        private List<InetSocketAddress> routers;
+        private long advertiseTimeout = 5000;
```
Comment (truncs) [inline tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java]: Why are these parameters configurable ?

File: tchannel-example/src/main/java/com/uber/tchannel/ping/PingRequestHandler.java [hunk 4]
```diff
@@ -27,12 +27,19 @@
...
-public class PingDefaultRequestHandler extends JSONRequestHandler<Ping, Pong> {
+public class PingRequestHandler extends JSONRequestHandler<Ping, Pong> {
 
+    @Override
     public Response<Pong> handleImpl(Request<Ping> request) {
-        return new Response.Builder<>(new Pong("pong!"), request.getEndpoint(), ResponseCode.OK)
-                .setHeaders(request.getHeaders())
-                .build();
+
+        return new Response.Builder<>(
```
Comment (truncs) [inline tchannel-example/src/main/java/com/uber/tchannel/ping/PingRequestHandler.java]: this style is more confusing than the previous ones, also can we have teh style changes in a separate diff

File: tchannel-example/src/main/java/com/uber/tchannel/hyperbahn/HyperbahnExample.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+import com.google.common.util.concurrent.ListenableFuture;
+import com.uber.tchannel.api.Response;
+import com.uber.tchannel.api.TChannel;
+import com.uber.tchannel.api.errors.TChannelError;
+import com.uber.tchannel.hyperbahn.api.HyperbahnClient;
+import com.uber.tchannel.hyperbahn.messages.AdvertiseResponse;
+import com.uber.tchannel.ping.PingRequestHandler;
+
+// Instructions:
```
Comment (truncs) [inline tchannel-example/src/main/java/com/uber/tchannel/hyperbahn/HyperbahnExample.java]: can this go in the README.md of this module ?

File: tchannel-core/src/main/java/com/uber/tchannel/schemes/JSONSerializer.java [hunk 6]
```diff
@@ -50,8 +50,12 @@ public String decodeEndpoint(ByteBuf arg1) {
     public Map<String, String> decodeHeaders(ByteBuf arg2) {
         String headerJSON = arg2.toString(CharsetUtil.UTF_8);
         arg2.release();
-        Map<String, String> headers = GSON.fromJson(headerJSON, HEADER_TYPE);
-        return (headers != null) ? headers : new HashMap<String, String>();
+        if (headerJSON == null || headerJSON.isEmpty() || headerJSON.equals("\"\"")) {
```
Comment (truncs) [inline tchannel-core/src/main/java/com/uber/tchannel/schemes/JSONSerializer.java]: that last conditions looks weird

File: tchannel-core/src/main/java/com/uber/tchannel/handlers/RequestRouter.java [hunk 7]
```diff
@@ -127,7 +129,14 @@ public void onSuccess(RawResponse response) {
             @Override
             public void onFailure(Throwable throwable) {
                 queuedRequests.decrementAndGet();
-                // TODO handle the failure case
+
+                // TODO better interface for sending errors
+                ErrorMessage error = new ErrorMessage(
+                        rawRequest.getId(),
+                        ErrorType.BadRequest,
```
Comment (truncs) [inline tchannel-core/src/main/java/com/uber/tchannel/handlers/RequestRouter.java]: is this really a bad request ?
The exception handling over here is different from the exception handling before in the same function.

File: tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java (hunk 1 above)
Comment (ShanniLi) [inline tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java]: AdvertiseInterval is fixed (1 min). Only retry on failure should back off. 
We don't really have retry on failure yet ...

File: tchannel-core/src/main/java/com/uber/tchannel/api/errors/TChannelError.java [hunk 8]
```diff
@@ -23,12 +23,15 @@
...
 
 public class TChannelError extends Exception {
+    public static final String ERROR_INTERRUPTED = "tchannel.interrupted";
+
     public static final String ERROR_INIT_TIMEOUT = "tchannel.connection.timeout";
+    public static final String ERROR_CONNECTION_FAILURE = "tchannel.socket";
 
     public final String type;
-    public final Exception subError;
+    public final Throwable subError;
```
Comment (junchaowu) [inline tchannel-core/src/main/java/com/uber/tchannel/api/errors/TChannelError.java]: no need to do this change.

File: tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java (hunk 2 above)
Comment (ShanniLi) [inline tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java]: The implementation where I picked up from @wjs doesn't have a daemon thread implementation. It is a blocking call waiting for advertisement to finish.

File: tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java (hunk 3 above)
Comment (ShanniLi) [inline tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java]: It is configuration in other languages. :)

Comment (anson627) [inline tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java]: final?

File: tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+        }
+
+        private static List<InetSocketAddress> loadRouters(String hostsFilePath) throws IOException {
+
+            List<String> hostPorts;
+            try (Reader reader = new FileReader(hostsFilePath)) {
+                hostPorts = new Gson().fromJson(reader, List.class);
+            }
+
+            List<InetSocketAddress> routers = new LinkedList<>();
```
Comment (anson627) [inline tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java]: ArrayList?

File: tchannel-core/src/main/java/com/uber/tchannel/handlers/RequestRouter.java (hunk 7 above)
Comment (ShanniLi) [inline tchannel-core/src/main/java/com/uber/tchannel/handlers/RequestRouter.java]: not sure which function you refer to, but, yes, this is the right response. Essentially, it cannot decode arg2 or arg3, it is quite a BadRequest.

File: tchannel-core/src/main/java/com/uber/tchannel/schemes/JSONSerializer.java (hunk 6 above)
Comment (ShanniLi) [inline tchannel-core/src/main/java/com/uber/tchannel/schemes/JSONSerializer.java]: agreed. :) That is what tchannel-node sends for empty headers.

File: tchannel-example/src/main/java/com/uber/tchannel/ping/PingRequestHandler.java (hunk 4 above)
Comment (ShanniLi) [inline tchannel-example/src/main/java/com/uber/tchannel/ping/PingRequestHandler.java]: yea, the old one looks better. I will make the change.

File: tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java (hunk 3 above)
Comment (ShanniLi) [inline tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java]: advertiseTimeout should be configurable

File: tchannel-example/src/main/java/com/uber/tchannel/hyperbahn/HyperbahnExample.java (hunk 5 above)
Comment (ShanniLi) [inline tchannel-example/src/main/java/com/uber/tchannel/hyperbahn/HyperbahnExample.java]: I feel it is useful for easy referencing to keep an instruction in the example file.

I will also add one to README.md

File: tchannel-core/src/main/java/com/uber/tchannel/api/errors/TChannelError.java (hunk 8 above)
Comment (ShanniLi) [inline tchannel-core/src/main/java/com/uber/tchannel/api/errors/TChannelError.java]: https://github.com/uber/tchannel-java/pull/87/files#diff-74e96f9ff03312330da9348801f6a552R119

File: tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java (hunk 9 above)
Comment (ShanniLi) [inline tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java]: good point, will fix.

File: tchannel-core/src/main/java/com/uber/tchannel/handlers/RequestRouter.java (hunk 7 above)
Comment (truncs) [inline tchannel-core/src/main/java/com/uber/tchannel/handlers/RequestRouter.java]: Not really ? Thats one of the case. Any exception thrown by handler would get reflected over here

File: tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java (hunk 2 above)
Comment (truncs) [inline tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java]: Using daemon threads would be better then !

File: tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+        responseFuture.addListener(new Runnable() {
+            @Override
+            public void run() {
+                if (destroyed.get()) {
+                    return;
+                }
+
+                scheduleAdvertise();
+            }
+        }, new Executor() {
```
Comment (truncs) [inline tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java]: starting a new executor is really expensive

**Comment by anson627** [comment]
lgtm

File: tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java (hunk 2 above)
Comment (ShanniLi) [inline tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java]: Do you have an example, e.g., reference to your original code. I can make the change.

File: tchannel-core/src/main/java/com/uber/tchannel/handlers/RequestRouter.java (hunk 7 above)
Comment (ShanniLi) [inline tchannel-core/src/main/java/com/uber/tchannel/handlers/RequestRouter.java]: Any exception happens at decoding should be considered as BadRequest (assuming decoding code logic is correct).
Anyway, the current implementation is much better than eating the error. So, I would consider the change a step forward.

File: tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java (hunk 10 above)
Comment (ShanniLi) [inline tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java]: Doesn't really matter when this ran once a minute.
Feel free to suggest a better style. I can make the update. Or you can send a PR for it. :)

**Comment by ShanniLi** [comment]
@truncs Need to move forward to unblock others. Feel free to create issues on me for improvements. I will be happy to fix them along the way.

**Pull request merged**